index.js

```diff
@@ -830,60 +805,108 @@ Tree.prototype._onToggle = function (d) {
+      ;(p._allChildren || (p._allChildren = [])).push(_n)
+    }
+  } else {
+    // Some type of root nodes. We treat those as expanded nodes
+    if (this.options.forest) {
+      this.root.splice(typeof idx === 'number' ? idx : this.root.length, 0, _n)
+    } else {
+      _n.collapsed = false
+      this.root = _n
+    }
+  }
+
+  return _n
+}
+
 /*
  * Adds a new node to the tree. Pass in d as the data that represents
  * the node, parent (which can be the parent object or an id), and an optional
  * index. If the index is sent, the node will be inserted at that index within the
  * parent's children.
+ *
+ * Receives optional `opts`, which supports:
+ *    `expand`: defaults to true, which will expand the incoming node's parent if the parent is selected
  */
-Tree.prototype.add = function (d, parent, idx) {
+Tree.prototype.add = function (d, parent, idx, opts = { expand: true }) {
   if (this._layout[d.id]) {
-    // can't add a node that we already have
+    // Already have this node, ignore
     return
   }
 
-  // internal node used for computing the layout
-  var _d = { id: d.id }
-
-  if (!parent && this.options.forest) {
-    // Forest tree and the new node is a new root
-    this.nodes[d.id] = d // Store the real node
-    this._layout[_d.id] = _d
-    this.root.splice(typeof idx === 'number' ? idx : this.root.length, 0, _d)
-    this._transitionWrap(this._slide)()
-    return d
-  } else if (parent) {
-    parent = this._layout[parent !== null && typeof parent === 'object' ? parent.id : parent]
-  } else if (!parent && !this.root) {
-    this.root = _d
-  } else {
-    // No parent, and not a new root node
-    return
-  }
+  // Turn parent into a parent object
+  parent = this._layout[parent !== null && typeof parent === 'object' ? parent.id : parent]
 
-  if (parent && parent.selected && parent.collapsed) {
+  if (opts.expand && parent && parent.selected && parent.collapsed) {
     // The parent is selected and collapsed, we want to expand its children (#259)
     delete parent.collapsed
     // Show the children immediately
     this._fly()
     this._forceRedraw()
   }
 
-  _d.parent = parent
-  this.nodes[d.id] = d
-  this._layout[_d.id] = _d
+  let _d = this._store(d, parent, idx)
+
+  this._transitionWrap(this._slide)()
 
-  if (typeof idx !== 'undefined') {
-    parent._allChildren.splice(idx, 0, _d)
-  } else if (parent) {
-    if (!parent._allChildren) {
-      parent._allChildren = []
+  return d
+}
+
+/*
+ * Adds multiple nodes to the tree. This is more efficient than invoking `.add` on each node
+ * as it only draws the nodes once, after the internal tree representation has been updated.
+ * Each node in the array should contain the arguments that should be passed to `.add`:
+ *  ```
+ *  {
+ *    data: // the node's data
+ *    parent: // optional parent of the node
+ *    idx: // idx to insert in the parent
+ *  }
+ *  ```
+ */
+Tree.prototype.addAll = function (nodes = []) {
```

Comment:
I don't really like this API signature. Any ideas?

Reply:
I don't see this being called anywhere (except the example), so I don't think I fully grasp the entire intended fix

Reply:
I'm also wondering if this fix will make it so we can go back to streaming the tree all the time rather than just sometimes since the bottleneck seemed to be unnecessarily drawing the tree several times.

We send the tree nodes in breadth-first manner.  Could it be made even faster if you just drew the tree once you had the first level of nodes, and then didn't draw it again unless it needed to expand somewhere? For non-forest trees, you would know you were on a level other than the first one once you see a second unique tree ID. For forest trees, you could do something similar

Reply:
This will be used when many orgs are added at once (like in a copy/paste that isn't a cube template). Right now we are calling `tree.add()` in a loop.

Reply:
I'm fine with the API signature; I'm not sure what would be better than `addAll()`. FWIW, that's what Java uses in its collection API.

It might be worth noting in the jsdoc that the new nodes will not be expanded. This is what we want, but it's different than the default behavior for `add()`

Reply:
@mattsgarlata This is a new method @jerryorr will hit from SB3. I also believe that's more or less how the tree works. When the tree is originally drawn, it only updates the DOM in the following situations:

1. It's drawing a root node.
1. It's drawing the selected node (based on initialSelection)
1. It's drawing a node and the node has an expanded parent.

All of these changes should be unrelated to the streaming stuff. We'll chat about that in a separate issue when I start investigating those org tree issues.

Reply:
Oh, I thought this was a fix for https://github.com/SpiderStrategies/Scoreboard/issues/22641, but I guess it was actually for https://github.com/SpiderStrategies/Scoreboard/issues/22662?

Another alternative could be to overload the `add` method, but it's probably clearer the way you have it here.